tests/testsuite/sbom.rs

```diff
@@ -0,0 +1,637 @@
+                "name": "dev",
+                "opt_level": "0",
+                "lto": "false",
+                "codegen_backend": null,
+                "codegen_units": null,
+                "debuginfo": 2,
+                "split_debuginfo": "{...}",
+                "debug_assertions": true,
+                "overflow_checks": true,
+                "rpath": false,
+                "incremental": false,
+                "panic": "unwind",
+                "strip": {
+                    "deferred": "None"
+                }
+            },
+            "packages": [],
+            "features": [],
+            "rustc": {
+                "version": "[..]",
+                "wrapper": null,
+                "workspace_wrapper": null,
+                "commit_hash": "[..]",
+                "host": "[..]",
+                "verbose_version": "{...}"
+            }
+        }
+        "#,
+    )
+}
+
+#[cargo_test]
+fn build_sbom_with_simple_build_script() {
+    let p = configured_project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+                build = "build.rs"
+            "#,
+        )
+        .file("src/main.rs", "#[cfg(foo)] fn main() {}")
+        .file(
+            "build.rs",
+            r#"fn main() {
+                println!("cargo::rustc-check-cfg=cfg(foo)");
+                println!("cargo::rustc-cfg=foo");
+            }"#,
+        )
+        .build();
+
+    p.cargo("build -Zsbom")
+        .masquerade_as_nightly_cargo(&["sbom"])
+        .run();
+
+    let path = with_sbom_suffix(&p.bin("foo"));
+    assert!(path.is_file());
+
+    assert_json_output(
+        path,
+        r#"
+        {
+            "format_version": 1,
+            "package_id": "path+file://[..]/foo#0.0.1",
+            "name": "foo",
+            "version": "0.0.1",
+            "source": "[ROOT]/foo",
+            "target": {
+                "kind": [
+                    "bin"
+                ],
+                "crate_types": [
+                    "bin"
+                ],
+                "name": "foo",
+                "edition": "2015"
+            },
+            "profile": {
+                "name": "dev",
+                "opt_level": "0",
+                "lto": "false",
+                "codegen_backend": null,
+                "codegen_units": null,
+                "debuginfo": 2,
+                "split_debuginfo": "{...}",
+                "debug_assertions": true,
+                "overflow_checks": true,
+                "rpath": false,
+                "incremental": false,
+                "panic": "unwind",
+                "strip": {
+                    "deferred": "None"
+                }
+            },
+            "packages": [
+                {
+                    "build_type": "build",
+                    "dependencies": [
+                        {
```

Comment:
I'm concerned with what happens when a crate is compiled two different ways under resolver v2. For example, once as a build dependency and once as a normal dependency. In this list of `dependencies` how would we tell the two entries apart?

Would it be better for the `dependencies` to be a sequence of indicies into the `packages` list?

Reply:
I'd also really like to see a test of this case.

Reply:
Hello @arlosi ,

I refactored the code that collects all packages & their dependencies. First all packages are collected, then for each package the indices of its dependencies are determined. It simplifies the serialization code slightly. Would like to get your feedback, if this is what you had in mind.